src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,258 @@
+        AdjustBackground(patch);
+        AdjustEnvironment(patch);
+        AddIceChunks(patch);
+        LogEvent(patch, totalTimePassed);
+    }
+
+    private void AdjustBackground(Patch patch)
+    {
+        patch.BiomeTemplate.Background = Background;
+        patch.CurrentSnapshot.Background = Background;
+        patch.BiomeTemplate.Sunlight.Colour = new Color(0.8f, 0.8f, 1, 1);
+        patch.CurrentSnapshot.LightColour = new Color(0.8f, 0.8f, 1, 1);
+    }
+
+    private void AdjustEnvironment(Patch patch)
+    {
+        var currentTemperature = patch.Biome.ChangeableCompounds[Compound.Temperature];
+        var currentSunlight = patch.Biome.ChangeableCompounds[Compound.Sunlight];
+
+        currentTemperature.Ambient = random.Next(-1, 5) - currentTemperature.Ambient;
+        currentSunlight.Ambient = 0.5f - currentSunlight.Ambient;
+
+        var changes = new Dictionary<Compound, float>
+        {
+            [Compound.Temperature] = currentTemperature.Ambient,
+            [Compound.Sunlight] = currentSunlight.Ambient,
+        };
+
+        patch.Biome.ApplyLongTermCompoundChanges(patch.BiomeTemplate, changes, new Dictionary<Compound, float>());
+    }
+
+    /// <summary>
+    ///   Gets chunks from Iceshelf patch template and applies them to the patches.
+    /// </summary>
+    private void AddIceChunks(Patch patch)
+    {
+        var templateBiome = SimulationParameters.Instance.GetBiome(TemplateBiomeForIceChunks);
+        foreach (var configuration in IceChunksConfigurations)
+        {
+            var iceChunkConfiguration = templateBiome.Conditions.Chunks[configuration];
+            iceChunkConfiguration.Density *= 10;
+            patch.Biome.Chunks.Add(Prefix + configuration, iceChunkConfiguration);
+        }
+    }
+
+    private void LogEvent(Patch patch, double totalTimePassed)
+    {
+        patch.LogEvent(new LocalizedString("GLOBAL_GLACIATION_EVENT"),
+            true, true, "GlobalGlaciationEvent.svg");
+
+        if (patch.Visibility == MapElementVisibility.Shown)
+        {
+            targetWorld.LogEvent(new LocalizedString("GLOBAL_GLACIATION_EVENT_LOG", patch.Name),
+                true, false, "GlobalGlaciationEvent.svg");
+        }
+
+        patch.AddPatchEventRecord(WorldEffectVisuals.GlobalGlaciation, totalTimePassed);
+    }
+
+    private void LogBeginningOfGlaciation()
+    {
+        targetWorld.LogEvent(new LocalizedString("GLOBAL_GLACIATION_START_EVENT_LOG"),
+            true, true, "GlobalGlaciationEvent.svg");
+    }
+
+    private void LogEndOfGlaciation()
+    {
+        targetWorld.LogEvent(new LocalizedString("GLOBAL_GLACIATION_END_EVENT_LOG"),
+            true, true, "GlobalGlaciationEvent.svg");
+    }
+
+    private void FinishEvent()
+    {
+        hasEventAlreadyHappened = true;
+        foreach (var index in modifiedPatchesIds)
+        {
+            if (!targetWorld.Map.Patches.TryGetValue(index, out var patch))
+                continue;
```

Comment:
Shouldn't this give a warning that a patch has exited the world as that's not expected to happen?